LunrCoreLmdb/LmdbIndex.cs

```diff
@@ -0,0 +1,329 @@
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            using var tx = Env.Value.BeginTransaction(TransactionBeginFlags.None);
+            using var db = tx.OpenDatabase(configuration: Config);
+
+            tx.Delete(db, KeyBuilder.BuildFieldVectorKey(key));
+            tx.Delete(db, KeyBuilder.BuildFieldVectorKeyKey(key));
+            return tx.Commit() == MDBResultCode.Success;
+        }
+
+		#endregion
+
+		#region Inverted Index Entries
+
+        public bool AddInvertedIndexEntry(string key, InvertedIndexEntry invertedIndexEntry, CancellationToken cancellationToken = default)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            using var tx = Env.Value.BeginTransaction(TransactionBeginFlags.None);
+            using var db = tx.OpenDatabase(configuration: Config);
+
+            tx.Put(db, KeyBuilder.BuildInvertedIndexEntryKey(key), invertedIndexEntry.Serialize(), PutOptions.NoDuplicateData);
+            tx.Put(db, KeyBuilder.BuildTokenSetWordKey(key), Encoding.UTF8.GetBytes(key), PutOptions.NoDuplicateData);
+
+            return tx.Commit() == MDBResultCode.Success;
+        }
+
+        public InvertedIndexEntry? GetInvertedIndexEntryByKey(string key, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            using var tx = Env.Value.BeginTransaction(TransactionBeginFlags.ReadOnly);
+            using var db = tx.OpenDatabase(configuration: Config);
+            using var cursor = tx.CreateCursor(db);
+
+            var sr = cursor.Set(KeyBuilder.BuildInvertedIndexEntryKey(key));
+            if (sr != MDBResultCode.Success)
+                return default;
+
+            var (r, _, v) = cursor.GetCurrent();
+            return r != MDBResultCode.Success ? default : v.AsSpan().DeserializeInvertedIndexEntry();
+        }
+
+        #endregion
+
+        #region TokenSets
+
+        public TokenSet IntersectTokenSets(TokenSet other, CancellationToken cancellationToken)
+        {
+            // FIXME: This is still reading fully into memory before intersection
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var builder = new TokenSet.Builder();
+
+            using var tx = Env.Value.BeginTransaction(TransactionBeginFlags.ReadOnly);
+            using var db = tx.OpenDatabase(configuration: Config);
+            using var cursor = tx.CreateCursor(db);
+
+            var sr = cursor.SetRange(KeyBuilder.BuildAllTokenSetWordKeys());
+            if (sr != MDBResultCode.Success)
+                return builder.Root;
+
+            var (r, _, v) = cursor.GetCurrent();
+			
+            while (r == MDBResultCode.Success && !cancellationToken.IsCancellationRequested)
+            {
+                var buffer = v.AsSpan().ToArray();
+                var word = Encoding.UTF8.GetString(buffer);
```

Comment:
Actually, you might want to move some of that to a helper method since it's done so many times.

Reply:
Sure thing, I added a helper to open read-only cursor ops.

[Update] In practice, no. I get `Invalid reuse of reader locktable slot` from lmdb. I see that my first pass was using `using` scopes which obviously won't work inside a method, but attempts to fix it didn't pan out. I'll have to revisit later.